Interaction regions for file and range inputs are incorrect

hortont424 opened the pull request:
#### 91f56dce7de3dc8cdecbc8c8fe355c8e6a9ca347
<pre>
Interaction regions for file and range inputs are incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=247645">https://bugs.webkit.org/show_bug.cgi?id=247645</a>

Reviewed by Aditya Keerthi.

* LayoutTests/interaction-region/input-type-file-region-expected.txt: Added.
* LayoutTests/interaction-region/input-type-file-region.html: Added.
* LayoutTests/interaction-region/input-type-range-region-expected.txt: Added.
* LayoutTests/interaction-region/input-type-range-region.html: Added.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
Responding to &quot;DOMActivate&quot; should count as responding to mouse-click,
like on macOS, because it is. I propagated this unnecessary behavior difference
forward in an earlier patch (251077@main), now it&apos;s time to unify the two platforms.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowElement):
(WebCore::shouldAllowNonPointerCursorForElement):
(WebCore::interactionRegionForRenderedRegion):
Disallow &lt;input type=&quot;file&quot;&gt; and &lt;input type=&quot;range&quot;&gt; from themselves generating
interaction regions; they defer to their inner button and thumb as the primary
interactible parts, respectively.

Ensure that the range thumb does generate a interaction region, even though
it uses the default cursor style, not a hand cursor.

* Source/WebCore/rendering/RenderFileUploadControl.cpp:
(WebCore::RenderFileUploadControl::paintObject):
(WebCore::RenderFileUploadControl::paintControl):
* Source/WebCore/rendering/RenderFileUploadControl.h:
RenderFileUploadControl previously did not generate *any* event regions of any
kind, because it aggressively cuts off painting if painting is disabled
(as it is for an event region painting pass). Call the parent class&apos;s implementation
of paintObject() even when painting is disabled, so it can emit event regions.

Canonical link: <a href="https://commits.webkit.org/256467@main">https://commits.webkit.org/256467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75e677e526548aa3b42ea8f9f3950c97ecf4af8c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95875 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5122 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105436 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5219 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/33884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88249 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/101267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101537 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/82482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/30877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39616 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/37298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4468 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/41461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/82482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43511 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39721 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -63,6 +65,37 @@ static CursorType cursorTypeForElement(Element& element)
+        return false;
+
+    if (is<HTMLInputElement>(element)) {
+        const HTMLInputElement& input = downcast<HTMLInputElement>(element);
```
`if (auto* input = dynamicDowncast<HTMLInputElement>(element))`

hortont424 replied:
true!

hortont424 committed:
Interaction regions for file and range inputs are incorrect
https://bugs.webkit.org/show_bug.cgi?id=247645

Reviewed by Aditya Keerthi.

* LayoutTests/interaction-region/input-type-file-region-expected.txt: Added.
* LayoutTests/interaction-region/input-type-file-region.html: Added.
* LayoutTests/interaction-region/input-type-range-region-expected.txt: Added.
* LayoutTests/interaction-region/input-type-range-region.html: Added.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
Responding to "DOMActivate" should count as responding to mouse-click,
like on macOS, because it is. I propagated this unnecessary behavior difference
forward in an earlier patch (251077@main), now it's time to unify the two platforms.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowElement):
(WebCore::shouldAllowNonPointerCursorForElement):
(WebCore::interactionRegionForRenderedRegion):
Disallow <input type="file"> and <input type="range"> from themselves generating
interaction regions; they defer to their inner button and thumb as the primary
interactible parts, respectively.

Ensure that the range thumb does generate a interaction region, even though
it uses the default cursor style, not a hand cursor.

* Source/WebCore/rendering/RenderFileUploadControl.cpp:
(WebCore::RenderFileUploadControl::paintObject):
(WebCore::RenderFileUploadControl::paintControl):
* Source/WebCore/rendering/RenderFileUploadControl.h:
RenderFileUploadControl previously did not generate *any* event regions of any
kind, because it aggressively cuts off painting if painting is disabled
(as it is for an event region painting pass). Call the parent class's implementation
of paintObject() even when painting is disabled, so it can emit event regions.

Canonical link: https://commits.webkit.org/256467@main
